A set of patches to evolve to a possible pre-release 1.5.0 pre.1

deathaxe opened the pull request:
I'd like to propose fixes for issues #74, #263, #340 and several enhancements to reduce disk usage and number of calls to `git show` / `git archive`.

I put all commits into this PR I would find worth to be included in the next release as they fix some of the most annoying issues.

For details about the changes please have a look onto the commit messages itself.

I hope to find some willing people to check the comprehensive changes thoroughly. I work with them for a while now and found them quite stable during the last days and weeks.

r-stein commented on git_gutter_settings.py:
```diff
@@ -178,6 +172,15 @@ def set_compare_against(self, git_tree, new_compare_against):
+        width = view.settings().get('git_gutter_show_in_minimap')
+        if width is None:
+            width = self.get('show_in_minimap', 1)
+        return width if 0 <= width <= 9 else 100000
```
You must also handle the values `True` and `False`, because they were valid and hence still in some settings files.

deathaxe replied:
I checked by setting to true and false in may user settings with success.

`False` means 0 and therefore hides gutter icons from minimap.
`True` means 1 and shows gutter icons with default width of 1px as before.

Did you encounter any functional issues with it?

r-stein replied:
Oh sorry for the late reply. Yes it does work without a problem. I still prefer something like `return int(width) if 0 <= width <= 9 else 100000` or just `width = int(width)` a line above.

kofalt commented:
I don't think I posted here, but I can confirm that this branch fixed #74 for me :+1: 

If you need any further testing, let me know!

deathaxe commented:
As this PR changes quite much, I would be pleased to any feedback about possible bugs in general. Are the gutter icons always updated as expected? Are there situations a modification in a text file or external status change of the file (stage, commit) is not recognized? Do you encounter any exceptions? Things like that would help. If no such strange things can be found and no one complains about code formating, I'd like to merge it to the master branch and ask to publish it as pre-release somewhat later, so more people can test it before final release.

kofalt commented:
To answer your questions:

* Are the gutter icons always updated as expected?

Yes, as far as I could tell. I even did some work over SSHFS and it was fine with a higher-latency file system. 

* Are there situations a modification in a text file or external status change of the file (stage, commit) is not recognized?

I could not find any situation where that happened - it does seem like there is a delay on initial file load (maybe it's the file opening, plus the 300ms debounce or what-not? maybe remove that specifically for file open?), but it's not very long.

*  Do you encounter any exceptions? 

Nope! And I've used it in repositories that do, and repos that do not, have smudge/clean filters. I have not checked on the other issues this PR closes.

deathaxe commented:
Thank you very much for your feedback. Sounds good.

You are right with the initial delay. It is caused mainly by the debounce delay which is applied to all events right now. I was already thinking about to use it for 'modified' events only, but as SublimeText sends events like 'activated', 'loaded' very quickly and several times after each other on file (re-)loads, I did not jet find a good solution for it without causing git_gutter to run too often.

I personally use `"debounce_delay": 500` to reduce latency a bit. The default is 1000ms.

kofalt commented:
That makes sense. I'm not afraid of thrashing I/O, so I might set that very low.

One thing worth testing is seeing what happens if the command takes longer than the debounce delay (maybe add a sleep to the command temporarily). Does a new update start processing even if an old one is still in flight? Etc.

deathaxe commented:
Thanks for your tip. There actually was nothing which prevented several evaluation runs at a time. 

I rebased all commits to the current master and squashed in all fixup! commits to not loose overview.

To avoid parallel runs I added a busy marker and ensure to limit debounce delay to a minimum on the other hand. On my way I decided to refactor the EventListener module, too, in order to catch the correct events.

Please give it another try. If no more issues occure, that's it.

kofalt commented:
Question: with your busy marker implementation, will it know to scan again after the in-flight one completes? If not, you could end up with stale data.

I'll try to give the latest version a test later today :slightly_smiling_face:

r-stein commented on README.md:
```diff
@@ -116,7 +116,17 @@ In addition the popup loads the css file `gitgutter_popup.css`, which can be mod
+ value  | description
+--------|-----------------
+ 0      | hide markers
+ 1..9   | width of markers
```
Is there any particular reason you have a fixed max width?

deathaxe replied:
The width is calculated in number of chars. There is no technical reason for limitation, but it 9px seemed a width of gutter markers, which still looked good. 15, 20 or even 30px cover too much text, which looked somehow ugly. But this is just a personal feeling to it.

r-stein replied:
I fully agree to this, but I still prefer to just allow the users to make it ugly. However you can do it as you want to :)

deathaxe replied:
Have no problem with it. Changed!

r-stein commented on README.md:
```diff
@@ -116,7 +116,17 @@ In addition the popup loads the css file `gitgutter_popup.css`, which can be mod
+--------|-----------------
+ 0      | hide markers
+ 1..9   | width of markers
+ <0, >9 | show full line
```
Better just write `-1` here

deathaxe replied:
ok.

r-stein commented:
I think this is distracting, because it is not directly clear that the overlapping arguments belong to the `min` function and are not Region args:

``` py
region = sublime.Region(start, min(
    end, start + self._minimap_size))
```
I would prefer something like this:
``` py
region = sublime.Region(
    start, min(end, start + self._minimap_size))
```
or (if line too long):
``` py
region = sublime.Region(
    start,
    min(end, start + self._minimap_size))
```
or (if line still too long):
``` py
region_end = min(end, start + self._minimap_size)
region = sublime.Region(start, region_end)
```

deathaxe commented:
@kofalt: As soon as an evaluation has finished, the next received event (modified,post-saved,activated) will trigger the next run. Means as soon as you continue writing, evaluation will start again. Did not encounter any issue with this strategy right now. I may add some kind of queue to start the next evaluation as soon as the old one finished, if any event was received in between, but didn't find it needed up to now.

kofalt commented:
Just an update: I've fallen pretty sick, so I won't be able to look at these new revisions. Might happen this weekend or something if I get better. But based on my previous experimenting it seemed to work well.
